src/geometry.rs

```diff
@@ -0,0 +1,188 @@
+    pub fn with_x(x: f32) -> Self {
+        Self::from_point_normal(&Vec3::new(x, 0.0, 0.0), &Vec3::new(1.0, 0.0, 0.0))
+    }
+
+    /// Create a `Plane` which is facing along the Y-Axis at the provided coordinate.
+    pub fn with_y(y: f32) -> Self {
+        Self::from_point_normal(&Vec3::new(0.0, y, 0.0), &Vec3::new(0.0, 1.0, 0.0))
+    }
+
+    /// Create a `Plane` which is facing along the Z-Axis at the provided coordinate.
+    pub fn with_z(z: f32) -> Self {
+        Self::from_point_normal(&Vec3::new(0.0, 0.0, z), &Vec3::new(0.0, 0.0, 1.0))
+    }
+
+    /// f32his `Plane` normal
+    pub fn normal(&self) -> &Vec3 {
+        &self.normal
+    }
+
+    /// Normalized representation of this `Plane`
+    pub fn normalize(&self) -> Self {
+        let distance = self.normal.mag();
+        Self {
+            normal: self.normal / distance,
+            bias: self.bias / distance,
+        }
+    }
+
+    /// Returns the dot product of this `Plane` and a provided `Vec3`
+    pub fn dot_point(&self, point: &Vec3) -> f32 {
+        self.normal.x * point.x + self.normal.y * point.y + self.normal.z * point.z + self.bias
+    }
+
+    /// Returns the dot product of this `Plane` and a provided `Vec3`
+    pub fn dot(&self, point: &Vec3) -> f32 {
+        self.normal.x * point.x + self.normal.y * point.y + self.normal.z * point.z
+    }
+
+    /// Returns the dot product of this `Plane` with another `Plane`
+    pub fn dot_plane(&self, plane: &Plane) -> f32 {
+        self.normal.x * plane.normal.x
+            + self.normal.y * plane.normal.y
+            + self.normal.z * plane.normal.z
+            + self.bias * plane.bias
+    }
+
+    /// Returns the intersection distance of the provided line given a point and direction, or `None` if none occurs.
+    pub fn intersect_line(&self, point: &Vec3, direction: &Vec3) -> Option<f32> {
+        let fv = self.dot(direction);
+        let distance = self.dot_point(point) / fv;
+        if fv.abs() > std::f32::MIN {
+            Some(distance)
+        } else {
+            None
+        }
+    }
+
+    /// Returns the intersection distance of the provided `Ray`, or `None` if none occurs.
+    pub fn intersect_ray(&self, ray: &Ray) -> Option<f32> {
+        self.intersect_line(&ray.origin, &ray.direction)
+    }
+}
+
+/// A Ray represents and infinite half-line starting at `origin` and going in specified unit length `direction`.
+#[derive(Debug, Copy, Clone)]
+pub struct Ray {
+    /// f32he origin point of the ray
+    pub origin: Vec3,
+    /// f32he normalized direction vector of the ray
+    pub direction: Vec3,
+}
+impl Ray {
+    /// Returns the distance along the ray which intersects with the provided `Plane`1
+    pub fn intersect_plane(&self, plane: &Plane) -> Option<f32> {
+        plane.intersect_ray(self)
+    }
+
+    /// Returns a `Vec3` along the ray at a distance `t` from it's origin.
+    pub fn at_distance(&self, z: f32) -> Vec3 {
+        self.origin - (self.direction * z)
```

Comment:
should be a plus, and can use mul_add
`self.direction.mul_add(Vec3::broadcast(z), self.origin)`